engines/sci/graphics/gfxdrivers.cpp

```diff
@@ -0,0 +1,508 @@
+		error("Failed to open '%s' or '%s'", _driverFiles[0], _driverFiles[1]);
+}
+
+SCI0_CGABWDriver::~SCI0_CGABWDriver() {
+	delete[] _monochromePatterns;
+}
+
+void SCI0_CGABWDriver::copyRectToScreen(const byte *src, int pitch, int x, int y, int w, int h) {
+	// We can't really properly fix the boundaries here, we have to do that before calling this function.
+	assert(!(w & _hAlign));
+	assert(!(x & _hAlign));
+
+	byte *dst1 = _compositeBuffer;
+	byte *dst2 = _compositeBuffer + (w << 1);
+	int ty = y & 7;
+	pitch -= w;
+
+	for (int i = 0; i < h; ++i) {
+		int tx = x & 3;
+		if (_earlyVersion) {
+			++ty;
+			for (int ii = 0; ii < (w >> 1); ++ii) {
+				uint16 p16 = reinterpret_cast<const uint16*>(_monochromePatterns)[((src[0] & 0x0f) << 4) | (src[1] & 0x0f)];
+				src += 2;
+				uint8 sh = (ty & 3) << 1;
+				uint8 lo = ((p16 & 0xff) >> sh) | ((p16 & 0xff) << (8 - sh));
+				uint8 hi = (p16 >> (8 + sh)) | ((p16 >> 8) << (8 - sh));
+				*dst1++ = *dst2++ = ((lo >> (6 - (tx << 1))) >> 1) & 1;
+				*dst1++ = *dst2++ = (lo >> (6 - (tx << 1))) & 1;
+				*dst1++ = *dst2++ = ((hi >> (4 - (tx << 1))) >> 1) & 1;
+				*dst1++ = *dst2++ = (hi >> (4 - (tx << 1))) & 1;
+				tx ^= 2;
+			}
+		} else {
+			for (int ii = 0; ii < w; ++ii) {
+				uint8 p = _monochromePatterns[((*src++ & 0x0f) << 3) + ty] >> (6 - (tx << 1));
+				*dst1++ = *dst2++ = (p >> 1) & 1;
+				*dst1++ = *dst2++ = p & 1;
+				tx = (tx + 1) & 3;
+			}
+			ty = (ty + 1) & 7;
+		}
+		src += pitch;
+		dst1 += (w << 1);
+		dst2 += (w << 1);
+	}
+
+	g_system->copyRectToScreen(_compositeBuffer, w << 1, x << 1, y << 1, w << 1, h << 1);
+}
+
+void SCI0_CGABWDriver::replaceCursor(const void *cursor, uint w, uint h, int hotspotX, int hotspotY, uint32 keycolor) {
+	// Instead of implementing the original cursor rendering code, we rely on the 8 bit cursor that
+	// has already been generated by the engine. Of course, that code could be moved to the EGA
+	// driver class and implemented again for each mode, but I don't see the benefit. Instead,
+	// we simply convert the colors as needed and scale the cursor...
+	assert(keycolor == 1);
+	keycolor = 0x0f;
+	w <<= 1;
+	const byte *s = reinterpret_cast<const uint8*>(cursor);
+	byte *d0 = _compositeBuffer;
+	byte *d1 = _compositeBuffer + w;
+
+	for (uint i = 0; i < h; ++i) {
+		for (uint ii = 0; ii < w; ++ii) {
+			*d0++ = *d1++ = *s ? (*s ^ 0x0e) : 0;
+			if (ii & 1)
+				++s;
+		}
+		d0 += w;
+		d1 += w;
+	}
+
+	CursorMan.replaceCursor(_compositeBuffer, w, h << 1, hotspotX << 1, hotspotY << 1, keycolor);
+}
+
+Common::Point SCI0_CGABWDriver::getMousePos() const {
+	Common::Point res = GfxDriver::getMousePos();
+	res.x >>= 1;
+	res.y >>= 1;
+	return res;
+}
+
+const char *SCI0_CGABWDriver::_driverFiles[2] = { "CGA320BW.DRV", "CGA320M.DRV" };
+
+SCI0_HerculesDriver::SCI0_HerculesDriver(int palIndex) : SCI0_DOSPreVGADriver(2, 720, 350, 0), _monochromePatterns(nullptr) {
```

Comment:
I usually use 350. I think we do that in SCUMM, too. The only reason is that we're more likely to get that screen format from the backend.
Same with CGA monochrome. Original resolution is 640x200, but that is too weird. So we use 640x400 and scale vertically.
The SCI Hercules driver renders in 640x300 and centers that on the screen. The original driver places the picture at y pos 24. I placed it at y pos 25 for that reason.

Reply:
I'm fairly sure that almost all backends that can handle high resolutions can also handle automatically resizing or centring the display within the nearest available hardware mode in order to handle games with unusual resolutions like Myst (which uses 544x332). The only backend I'm not sure about is Atari, but hopefully @mikrosk can confirm this.

> The SCI Hercules driver renders in 640x300 and centers that on the screen. The original driver places the picture at y pos 24. I placed it at y pos 25 for that reason.

I wonder if it would be useful to have an optional cropped mode for use with small screens or perfect pixel stretch modes?

> Same with CGA monochrome. Original resolution is 640x200, but that is too weird. So we use 640x400 and scale vertically.

I'll make a note on the issue tracker for the future about this. I'd been wanting to do something to handle scaling 640x256 for an external engine, but it looks like it could be useful for quite a few existing engines.

Reply:
> I'm fairly sure that almost all backends that can handle high resolutions can also handle automatically resizing or centring the display within the nearest available hardware mode in order to handle games with unusual resolutions like Myst (which uses 544x332). The only backend I'm not sure about is Atari, but hopefully @mikrosk can confirm this.
> 
> > The SCI Hercules driver renders in 640x300 and centers that on the screen. The original driver places the picture at y pos 24. I placed it at y pos 25 for that reason.
> 
> I wonder if it would be useful to have an optional cropped mode for use with small screens or perfect pixel stretch modes?

Like just setting the screen to 640x300 and getting rid of the centering offsets?

Reply:
> Like just setting the screen to 640x300 and getting rid of the centering offsets?

Essentially, just with a GUI option to toggle the behaviour.

Reply:
>  The only backend I'm not sure about is Atari, but hopefully @mikrosk can confirm this.

Centering is fine on Atari so yes, Myst & friends work well. However for the resolutions in this PR I have two notes:

- On VGA displays (which most users use these days), the maximum allowed resolution is 640x480. Only RGB/TV output can go (and the backend supports it) up to 768x480 so the usage of Hercules resolution will be limited (I'll just mention it in the README)

-  as hinted above, it may not be clear how to interpret CGA resolution: on Atari, we have the option to actually use native 640x200 (!) so I'd definitely love to use that; currently there's only one similar mechanism, the aspect ratio correction, which is controlled by a flag (and again, is supported by hardware on Atari) ... so some hypothetical flag for CGA games like "this game's resolution is expected to be shown in double height" would be useful

Reply:
OK, I've opened an issue: https://bugs.scummvm.org/ticket/15239

Reply:
Btw, the wikipedia article is quite clear on the 348 vs 350 difference:

- 80 × 25 text mode with 9 × 14 pixel font (effective resolution of 720 × 350, MDA-compatible)
- 720 × 348 graphics mode pixel-addressable graphics

So yes, ScummVM should use 348.

Reply:
Like I already said, the choice of 350 is not based on incorrect assumptions about the original hardware. It is just a way to better cope with the backends (which are not all up to the same standards). Same reason e. g. why I use 320 x 240 for EOB Sega CD and not 320 x 224.

Reply:
I'm fairly certain that all of our current backends can handle any resolution, and that the main difference is if the backend  centres the display within a larger resolution (SDL 1.2, Atari, N64), scales the picture using the graphics hardware (OpenGL, SDL2, 3DS, DC, PSP, Wii) or defers to a third-party frontend (Libretro). Since Myst has made use of that functionality for a long time, I think it's safe to say that any case where this doesn't work should be fixed in the backend rather than in the engines.

I'm also not sure if 350 is more or less standard than 348 is on modern devices.